### PR TITLE
Cancel rolling upgrades before deleting VMSS & extensions

### PR DESCRIPTION
**Pull request by harshavmb**
This is redo work of [PR-15634](https://github.com/hashicorp/terraform-provider-azurerm/pull/15634). The old PR was having conflicts & expects an additional argument to cancel before deleting VMSS & extensions. 

Since `destroy` call is a clear intention of deleting VMSS & associated resources like extensions, rolling upgrades could be cancelled as they are trivial.

As of now, we write destroy time provisioners to cancel rolling upgrades which are error prone. We hope this will improve the way VMSS & extensions are deleted in a clean manner. 

Our problem statement had been explained to Gauthier Donikian, Hashicorp few months ago. We expect some feedback from Hashicorp team. 

Thanks,
Harsha

### CODE REVIEW

**Comment by katbyte** [comment]
thanks for re-opening this @harshavmb - its on our list to review soon - as this is supersedes #15634 i'm going to close that one out

**Comment by harshavmb** [comment]
Hi @katbyte @tombuildsstuff,

I've got another idea today to take out the `rolling_upgrade_policy` outside of `azurerm_linux_virtual_machine_scale_set`. This way, there is no need of writing explicit cancel rolling upgrade calls made to ARM. 

Are you okay with this approach?

**Comment by harshavmb** [comment]
Hi @katbyte / @tombuildsstuff ,

Any update on this PR? We need some update on this before we end up with some code conflicts. The last PR (already closed) was also left un-reviewed for longtime.

**Comment by evaroquaux** [comment]
Hello,
Any update on this contribution?
Still highly expected, for a long time...
Thank you

**Comment by harshavmb** [comment]
Hi @stephybun ,

Your feedback is highly appreciated. We have been managing all these times with ugly `destroy` provisioners. If you think this PR helps, please share your feedback as this has been in your review queue for many months.

File: internal/services/compute/client/helper.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"strings"
+)
+
+func (c *Client) CancelRollingUpgradesBeforeDeletion(ctx context.Context, resourceGroupName string, vmScaleSetName string) error {
+	future, err := c.VMScaleSetRollingUpgradesClient.Cancel(ctx, resourceGroupName, vmScaleSetName)
+
+	// If rolling upgrades haven't been run (when VMSS are just provisioned with rolling upgrades but no extensions, auto-scaling are run )
+	// we can not cancel rolling upgrades
+	// API call :: GET https://management.azure.com/subscriptions/{subId}/resourceGroups/{rgName}/providers/Microsoft.Compute/virtualMachineScaleSets/{vmSSName}/rollingUpgrades/latest?api-version=2021-07-01
+	// Azure API throws 409 conflict error saying "The entity was not found in this Azure location."
```
Comment (stephybun) [inline internal/services/compute/client/helper.go:17]: @harshavmb I'm trying to reproduce this particular behaviour but am unable to. Could you provide some more details on what you mean by auto-scaling? As far as I can tell the resource doesn't expose any auto scaling properties.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @harshavmb and apologies being unable to get to this sooner.

I think this behaviour makes sense and is something we can incorporate into the resource. I've left suggestions which will bring this functionality more in-line with the rest of the provider. These will need to be made before we can consider merging this.

We also need to add a test for all three resources, this is a bit tricky since we need to simulate the behaviour of deleting the VMSS when rolling upgrades are happening in an automated fashion. It isn't blatantly obvious how to do that unless you're very familiar with our testing framework which is why I've gone ahead and written one, unfortunately I'm unable to push it to your branch since I lack permission.

I have the test and all of the suggested changes on my branch, which in the interest of time I could open as a separate PR to supersede this one. Alternatively I can provide the test and test config to you in another way. Let me know which you prefer.

**Comment by harshavmb** [comment]
> Thanks for this PR @harshavmb and apologies being unable to get to this sooner.
> 
> I think this behaviour makes sense and is something we can incorporate into the resource. I've left suggestions which will bring this functionality more in-line with the rest of the provider. These will need to be made before we can consider merging this.
> 
> We also need to add a test for all three resources, this is a bit tricky since we need to simulate the behaviour of deleting the VMSS when rolling upgrades are happening in an automated fashion. It isn't blatantly obvious how to do that unless you're very familiar with our testing framework which is why I've gone ahead and written one, unfortunately I'm unable to push it to your branch since I lack permission.
> 
> I have the test and all of the suggested changes on my branch, which in the interest of time I could open as a separate PR to supersede this one. Alternatively I can provide the test and test config to you in another way. Let me know which you prefer.

Hi @stephybun ,

Many Thanks for your review. If you have all the changes you needed to push from your branch, I'm fine with it. This PR can be closed when you create the PR. 

Also, I must admit I can't write the three test cases you mentioned. 

Thanks again for your help. Very much appreciated. 

Thanks,
Harsha

**Comment by stephybun** [comment]
Closing in favour of #22991

**Pull request closed without merging**